### PR TITLE
CLOUDP-195419: Improve Cloud Provider Access reconciliation flow

### PR DESCRIPTION
**Pull request by helderjs**
### All Submissions:

* [x] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if there is one).
* [ ] Update docs/release-notes/release-notes.md if your changes should be included in the release notes for the next release.

### CODE REVIEW

**Review by igor-karpukhin — APPROVED** [review]
Great work! 👍

File: pkg/controller/atlasproject/cloud_provider_access_test.go [hunk 1]
```diff
@@ -26,30 +31,953 @@ func (c *cloudProviderAccessClient) GetRole(_ context.Context, _ string, _ strin
...
+					return &mongodbatlas.Response{}, nil
+				},
+			},
+		}
+		workflowCtx := &workflow.Context{
+			Client: atlasClient,
+		}
+
+		result, err := syncCloudProviderAccess(context.TODO(), workflowCtx, "projectID", cpas)
+		assert.NoError(t, err)
```
Comment (josvazg) [inline pkg/controller/atlasproject/cloud_provider_access_test.go:135]: nit

Should this be `require.NoError(...)` instead? Or is the nil error also part of the final status expected in a way?

**Review by josvazg — COMMENTED** [review]
LGTM

Great and extensive unit tests!
This builds up more confidence that our code actually does what it "says on the tin".

**Review by josvazg — APPROVED** [review]

File: pkg/controller/atlasproject/cloud_provider_access_test.go (hunk 1 above)
Comment (helderjs) [reply pkg/controller/atlasproject/cloud_provider_access_test.go:135]: As `require` fails the test immediately it's a bit annoying not knowing what went wrong in the first run. Given that, I think makes more sense to use it in places where we know, "if this doesn't work, everything else won't work and we should stop asap"

**Review by roothorp — APPROVED** [review]
😍

**Pull request merged**